baseplate/sidecars/publisher_queue_utils.py

```diff
@@ -0,0 +1,104 @@
+from baseplate.lib.message_queue import RemoteMessageQueue
+from baseplate.lib.message_queue import TimedOutError
+from baseplate.server import make_listener
+from baseplate.server.thrift import make_server
+from baseplate.thrift.message_queue import RemoteMessageQueueService
+from baseplate.thrift.message_queue.ttypes import CreateResponse
+from baseplate.thrift.message_queue.ttypes import GetResponse
+from baseplate.thrift.message_queue.ttypes import PutResponse
+from baseplate.thrift.message_queue.ttypes import ThriftTimedOutError
+
+
+class RemoteMessageQueueHandler:
+    """Create an InMemoryMessageQueue locally and expose get/put methods.
+
+    See this overview for more details:
+    https://docs.google.com/document/d/1soN0UP9P12u3ByUwH_t47Uw9GwdVZRDuRFT0MvR52Dk/
+
+    """
+
+    def __init__(self) -> None:
+        # Store the queue by name with its max messages
+        self.queues: Dict[str, MessageQueue] = {}
+
+    def create_queue(self, queue_name: str, max_messages: int) -> CreateResponse:
+        queue = InMemoryMessageQueue(max_messages)
+        self.queues[queue_name] = queue
+
+        return CreateResponse()
+
+    def get(self, queue_name: str, timeout: Optional[float] = None) -> GetResponse:
+        try:
+            queue = self.queues[queue_name]
+            # Get element from list, waiting if necessary
+            result = queue.get(timeout)
+        # If the queue timed out, raise a timeout as the server response
+        except TimedOutError as e:
+            raise ThriftTimedOutError from e
+
+        return GetResponse(result)
+
+    def put(self, queue_name: str, message: bytes, timeout: Optional[float] = None) -> PutResponse:
+        try:
+            queue = self.queues[queue_name]
+            queue.put(message, timeout)
+        except TimedOutError as e:
+            raise ThriftTimedOutError from e
+        return PutResponse()
+
+
+@contextlib.contextmanager
+def start_queue_server(host: str, port: int) -> Generator[StreamServer, None, None]:
+    # Start a thrift server that will store the queue data in memory
+    processor = RemoteMessageQueueService.Processor(RemoteMessageQueueHandler())
+    server_bind_endpoint = config.Endpoint(f"{host}:{port}")
+    listener = make_listener(server_bind_endpoint)
+    server = make_server(server_config={}, listener=listener, app=processor)
+
+    # run the server until our caller is done with it
+    server_greenlet = gevent.spawn(server.serve_forever)
+    try:
+        yield server
+    finally:
+        server_greenlet.kill()
+
+
+def create_queue(
+    queue_type: QueueType,
+    queue_full_name: str,
+    max_queue_size: int,
+    max_element_size: int,
+    host: str = DEFAULT_QUEUE_HOST,
+    port: int = DEFAULT_QUEUE_PORT,
+) -> MessageQueue:
+    # See this overview for the relationship between InMemoryMessageQueues & RemoteMessageQueues
+    # https://docs.google.com/document/d/1soN0UP9P12u3ByUwH_t47Uw9GwdVZRDuRFT0MvR52Dk/
```

Comment:
this is public code, but references an internal doc -- we should add the documentation to this repo